TLS 1.3 signature algorithm and key selection

ryru opened the pull request:
This PR addresses:

- Flaw in reading signature algorithm and supported groups extension #25 
- Enhanced key selection in TLS 1.3 based on client signature algorithms #16 
- Improve signature parameter handling #11

Commit:
tls-server: Fix invalid signature algorithm and supported groups parsing

The extension content length field was wrongly add to the content data.

Fixes: 06109c471785e7bf09487e31c0a71cb325cc7926 ("Implemented "signature
algorithm" hello extension")

Commit:
tls-crypto: Add signature scheme constants

Some peers, e.g. Firefox, send a wide range of signature algorithms. To
prevent numeric identifiers in the log these algorithms are added here.

Commit:
tls-server: Rework TLS key and cipher suite selection

based on client offered signature algorithms.

Commit:
tls-crypto: Distinguish between signing and verifying signature schemes

strongSwan supports RSA_PSS_RSAE schemes for signing but does not
differentiate between rsaEncryption and rsassaPss encoding. Thus
RSA_PSS_PSS schemes are only used for verifying signatures.

ryru commented on src/libtls/tls_server.c:
```diff
@@ -189,24 +192,51 @@ static bool select_suite_and_key(private_tls_server_t *this,
 		this->suite = this->crypto->select_cipher_suite(this->crypto, suites,
-												  		count,
-												  		key->get_type(key));
+												  		count, type);
```
I think the above approach could also be used in legacy TLS condition branch, although the logic is a bit different. Also I think I would have to consider EdDSA of [RFC 8422](https://tools.ietf.org/html/rfc8422) you mentioned in #16, right?

ryru commented on src/libtls/tls_crypto.c:
```diff
@@ -2426,3 +2426,51 @@ tls_named_group_t tls_ec_group_to_curve(diffie_hellman_group_t group)
+	{
+		int i;
+
+		for (i = 0; i < countof(schemes); i++)
```
I extracted this functionality from `tls_server.c` to `tls_crypto.c` because of the existing `tls_signature_scheme_t`-to-`signature_params_t` mapping in `schemes`.

ryru commented on src/libtls/tls_crypto.c:
```diff
@@ -2426,3 +2426,51 @@ tls_named_group_t tls_ec_group_to_curve(diffie_hellman_group_t group)
+									tls_version_t max_version)
+{
+	bio_reader_t *reader;
+	array_t *types;
```
Maybe an array is a bit of an overkill here... I use it to mimic a Set ADT that only contains each available key type once.

I think it would be helpful to have a set-like ADT in strongSwan or enhance `array.c` with a `contains` method. This would reduce my nested loops. I am absolutely open to other approaches though.

tobiasbrunner reviewed (changes requested):
As discussed in the telco.

Commit:
tls-server: Check if peer sent hash and signature algorithms

It is mandatory to receive the `signature_algorithms` extension when a
server is authenticating over certificates, which is always the case.

ryru commented on src/libtls/tls_crypto.c:
```diff
@@ -2425,3 +2425,60 @@ tls_named_group_t tls_ec_group_to_curve(diffie_hellman_group_t group)
+	int i;
+
+	ht = hashtable_create((hashtable_hash_t) hashtable_hash_ptr,
+					   	  (hashtable_equals_t) hashtable_equals_ptr, 4);
```
Not sure if this is the correct way to create hashes and compare them...

tobiasbrunner replied:
No, this compares the pointers, not the values they point to. So you better write your own compare/hash functions (the latter usually use `chunk_hash[_inc]()`). You should find plenty of examples in other places.

ryru commented on src/libtls/tls_crypto.c:
```diff
@@ -2425,3 +2425,60 @@ tls_named_group_t tls_ec_group_to_curve(diffie_hellman_group_t group)
+			key_type_t type;
+
+			type = key_type_from_signature_scheme(schemes[i].params.scheme);
+			ht->put(ht, (void*)type, (void*)type);
```
Same here, not sure about that.

tobiasbrunner replied:
No need to cast to `void*`.

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -351,21 +353,38 @@ static status_t process_client_hello(private_tls_server_t *this,
 		switch (extension_type)
 		{
 			case TLS_EXT_SIGNATURE_ALGORITHMS:
+				if (!extension->read_data16(extension, &content))
```
Why not reuse `extension_data`?

tobiasbrunner commented on src/libtls/tls_crypto.c:
```diff
@@ -277,22 +277,24 @@ ENUM_NEXT(tls_signature_scheme_names,
 ENUM_NEXT(tls_signature_scheme_names,
-		  TLS_SIG_RSA_PKCS1_SHA256, TLS_SIG_RSA_PKCS1_SHA256, TLS_SIG_ECDSA_SHA1,
-	"RSA_PKCS1_SHA256");
+		  TLS_SIG_RSA_PKCS1_SHA224, TLS_SIG_ECDSA_SHA224, TLS_SIG_ECDSA_SHA1,
```
It doesn't seem possible to add a comment to a specific commit (message), so I'm adding this here. It would probably make sense to write something about the reason for adding these (e.g. that Firefox sends them and this avoids numeric identifiers in the log).

ryru replied:
Why do you think? Is it to clarify that not all of the added constants are usable/configurable?

I would suggest one comment just before `ENUM_BEGIN` rather than the same comment multiple times above all added signature constants. Do you agree?

```
/*
* Some peers, e.g. Firefox, send a wide range of signature algorithms. To prevent numeric identifiers in the log these algorithms are added here.
* 
* - RSA_PKCS1_SHA224
* - DSA_SHA224
* - ECDSA_SHA224
* - RSA_PKCS1_SHA256
* - DSA_SHA256
* - RSA_PKCS1_SHA384
* - DSA_SHA384
* - RSA_PKCS1_SHA512
* - DSA_SHA512
*/
```

tobiasbrunner replied:
I don't think a comment is necessary. It would just help to see in the history why these old identifiers were added after such a long time.

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -408,6 +408,14 @@ static status_t process_client_hello(private_tls_server_t *this,
 	}
 	extensions->destroy(extensions);
 
+	if (!this->hashsig.len)
```
Is hat true for all TLS versions?

ryru replied:
You are right!

TLS 1.2 states in 7.4.1.4.1:
```
   If the client does not send the signature_algorithms extension, the
   server MUST do the following:

   -  If the negotiated key exchange algorithm is one of (RSA, DHE_RSA,
      DH_RSA, RSA_PSK, ECDH_RSA, ECDHE_RSA), behave as if client had
      sent the value {sha1,rsa}.

   -  If the negotiated key exchange algorithm is one of (DHE_DSS,
      DH_DSS), behave as if the client had sent the value {sha1,dsa}.

   -  If the negotiated key exchange algorithm is one of (ECDH_ECDSA,
      ECDHE_ECDSA), behave as if the client had sent value {sha1,ecdsa}.

   Note: this is a change from TLS 1.1 where there are no explicit
   rules, but as a practical matter one can assume that the peer
   supports MD5 and SHA-1.
```

TLS 1.3 states in 4.2.3:
```
If a server is authenticating via
   a certificate and the client has not sent a "signature_algorithms"
   extension, then the server MUST abort the handshake with a
   "missing_extension" alert (see Section 9.2).
```

I'll make the alert conditional for TLS 1.3 only.

tobiasbrunner commented on src/libtls/tls_crypto.c:
```diff
@@ -2425,3 +2425,68 @@ tls_named_group_t tls_ec_group_to_curve(diffie_hellman_group_t group)
+	int i;
+
+	ht = hashtable_create((hashtable_hash_t) hash_key_type,
+					   	  (hashtable_equals_t) equals_key_type, 4);
```
Don't add a space after the cast. On the second line there is a whitespace error before the cast.

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -167,71 +168,155 @@ struct private_tls_server_t {
+					this->server_auth->destroy(this->server_auth);
+					this->server_auth = auth_cfg_create();
+					key = lib->credmgr->get_private(lib->credmgr, type,
+									 				this->server,
```
Whitespace error.

tobiasbrunner commented on src/libtls/tls_crypto.c:
```diff
@@ -2425,3 +2425,68 @@ tls_named_group_t tls_ec_group_to_curve(diffie_hellman_group_t group)
+		{
+			key_type_t tmp;
+
+			tmp= key_type_from_signature_scheme(schemes[i].params.scheme);
```
I'd declare these variables (also `type` below) at the start of the function (same type, results in smaller blocks). And I usually call the non-pointer one I do hashtable lookups with `lookup` (`tmp` is a bit generic). Also a space is missing here.

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -21,6 +21,7 @@
 #include <utils/debug.h>
 #include <credentials/certificates/x509.h>
 #include <collections/array.h>
+#include <collections/hashtable.h>
```
Should get pulled in via `tls_crypto.h` anyway.

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -167,71 +168,155 @@ struct private_tls_server_t {
+/**
+ * Create a hashtable of an intersection of server and peer supported key types
+ */
+static hashtable_t *create_common_key_types(chunk_t hashsig,
```
Maybe return an array (`sizeof(key_type_t)`)? Would provide a simpler interface for the caller and no need to free the removed elements (could simply use `while (array_remove(key_types, ARRAY_HEAD, &type))`). Creating a hashtable from `hashsig` before looping could also be an option to avoid having to parse and map it multiple times.

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -167,71 +168,155 @@ struct private_tls_server_t {
+	key_types = create_common_key_types(this->hashsig, version_min, version_max);
+	if (!key_types)
+	{
+		DBG1(DBG_TLS, "no common signature algorithm found");
```
`algorithm` -> `algorithms`. The number of elements should probably be checked as this always returns a valid pointer it seems.

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -167,71 +168,155 @@ struct private_tls_server_t {
 	}
+	if (!this->suite || !key)
+	{
+		DBG1(DBG_TLS, "received cipher suites unacceptable");
```
Perhaps also mention the signature schemes.

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -167,71 +168,155 @@ struct private_tls_server_t {
+				key_types->remove_at(key_types, enumerator);
+				free(v);
+
+				if (!key)
```
This should always be the case at this point (there is a `!key` check above that returns from the function). To keep the order the same, just free the existing key (and I guess auth_cfg), do the lookup with the new type (continue if none is found) and then look for a suite. Then go to the next type and free the key again etc. If you use an array you could even get rid of the `!this->suite` check and just do `while (!this->suite && array_remove...)` to check another key type.

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -167,71 +168,155 @@ struct private_tls_server_t {
+		key_types->destroy(key_types);
+		return FALSE;
+	}
+	enumerator = key_types->create_enumerator(key_types);
+	while (enumerator->enumerate(enumerator, &k, &v))
+	{
+		key_types->remove_at(key_types, enumerator);
+		free(v);
+	}
+	enumerator->destroy(enumerator);
+	key_types->destroy(key_types);
```
You could use `destroy_function()` and pass `free()` as callback. But using an array will be simpler.

tobiasbrunner commented on src/libtls/tls_crypto.h:
```diff
@@ -691,4 +692,22 @@ int tls_crypto_get_supported_groups(diffie_hellman_group_t **groups);
+ * @param max_version	maximum negotiated TLS version
+ * @return				hashtable of key types
+ */
+hashtable_t *tls_get_supported_key_types(tls_version_t min_version,
```
Since you are only enumerating this once now and don't actually use it as hashtable, it would probably be nicer to return an `enumerator_t` that hides the fact that a hashtable was used to build this list (also doesn't add a dependency to hashtable.h in the header). See the implementation for a suggestion how to implement it.

tobiasbrunner commented on src/libtls/tls_crypto.c:
```diff
@@ -2425,3 +2425,65 @@ tls_named_group_t tls_ec_group_to_curve(diffie_hellman_group_t group)
+			}
+		}
+	}
+	return ht;
```
Instead of returning the hashtable directly, you could wrap the hashtable's enumerator in a filter:

```suggestion
	return enumerator_create_filter(ht->create_enumerator(ht),
									filter_key_types, ht, destroy_key_types);
```
With two functions like this:

```c

CALLBACK(filter_key_types, bool,
	void *data, enumerator_t *orig, va_list args)
{
	key_type_t *key_type, *out;

	VA_ARGS_VGET(args, out);

	if (orig->enumerate(orig, NULL, &key_type))
	{
		*out = *key_type;
		return TRUE;
	}
	return FALSE;
}

CALLBACK(destroy_key_types, void,
	hashtable_t *ht)
{
	ht->destroy_function(ht, (void*)free);
}
```

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -167,71 +167,120 @@ struct private_tls_server_t {
+										tls_version_t version_min,
+										tls_version_t version_max)
+{
+	hashtable_t *ht;
+	array_t *key_types;
+	enumerator_t *enumerator;
+	int *k;
+	key_type_t *v, lookup;
+	uint16_t sig_scheme;
+
+	ht = tls_get_supported_key_types(version_min, version_max);
+	if (!ht)
+	{
+		return NULL;
+	}
+	key_types = array_create(sizeof(key_type_t), 1);
+	enumerator = ht->create_enumerator(ht);
+	while (enumerator->enumerate(enumerator, &k, &v))
```
Without hashtable, you could simplify this to the following (also, even before the change to enumerator_t, the function always returned a valid pointer):

```suggestion
	array_t *key_types;
	enumerator_t *enumerator;
	key_type_t v, lookup;
	uint16_t sig_scheme;

	key_types = array_create(sizeof(key_type_t), 8);
	enumerator = tls_get_supported_key_types(version_min, version_max);
	while (enumerator->enumerate(enumerator, &v))

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -167,71 +167,120 @@ struct private_tls_server_t {
+		reader->destroy(reader);
+	}
+	enumerator->destroy(enumerator);
+	ht->destroy_function(ht, (void*)free);
```
This is wrapped by the enumerator

```suggestion

```

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -167,71 +167,120 @@ struct private_tls_server_t {
+			   reader->read_uint16(reader, &sig_scheme))
+		{
+			lookup = tls_signature_scheme_to_key_type(sig_scheme);
+			if (*v == lookup)
```
```suggestion
			if (v == lookup)
```

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -167,71 +167,120 @@ struct private_tls_server_t {
+	version_min = this->tls->get_version_min(this->tls);
+	version_max = this->tls->get_version_max(this->tls);
+	key_types = create_common_key_types(this->hashsig, version_min, version_max);
+	if (!key_types || !array_count(key_types))
```
The function always returns an array (even was the case before as a hashtable was always returned).

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -167,71 +167,120 @@ struct private_tls_server_t {
-			}
-			key->destroy(key);
-
+		if (type)
```
We have a key at this point, so `type` must be defined as it was used in that lookup.

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -167,71 +167,120 @@ struct private_tls_server_t {
-			}
+		}
+		while (!this->suite && array_remove(key_types, ARRAY_HEAD, &type))
+		{	/* find a cipher suite that supports the key */
```
```suggestion
		{	/* find a key and cipher suite for one of the remaining key types */
```

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -167,71 +167,120 @@ struct private_tls_server_t {
+		}
+		while (!this->suite && array_remove(key_types, ARRAY_HEAD, &type))
+		{	/* find a cipher suite that supports the key */
+			key->destroy(key);
```
If we find no key for a type, this won't be defined anymore.
```suggestion
			DESTROY_IF(key);
```

tobiasbrunner commented on src/libtls/tls_server.c:
```diff
@@ -167,71 +167,120 @@ struct private_tls_server_t {
 			this->server_auth = auth_cfg_create();
 			key = lib->credmgr->get_private(lib->credmgr, type, this->server,
 											this->server_auth);
 			if (!key)
```
You could change this to `if (key)` and replace `continue;` with `this->suite = ...;`.

tobiasbrunner reviewed (changes requested):
Looks good, but I still found some things, sorry :)

ryru commented:
Thank you very much for the review. That is absolutely fine! :-)

Callback `filter_key_types` shrinks the enumerator to `key_type_t`. This simplifies enumeration for the caller. But the TLS version range filtering still needs to be done in `tls_get_supported_key_types`, because we can not pass the values to the callback filter. Is my understanding of the enumerator callbacks approach correct?

tobiasbrunner commented:
> This simplifies enumeration for the caller. But the TLS version range filtering still needs to be done in `tls_get_supported_key_types`, because we can not pass the values to the callback filter.

While you could theoretically pass them to `enumerate()` to receive them in in the filter this is not ideal because you'd need access to the schemes to filter them (or you'd have to store for which versions a certain key type is valid). And you'd have to pass them as pointers, which is not that nice. So it's definitely easier to just filter them in the constructor of the enumerator.

tobiasbrunner reviewed (approved):
Looks good.

Just as an additional comment regarding your question about the enumerator above. It would also have been possible to create a more elaborate enumerator (instead of simply wrapping another) that does the actual filtering while enumerating. The private struct of that enumerator would have to store the versions passed to the constructor, the index into the schemes array and the hashtable to avoid returning duplicate key types, then most of the code in the constructor could be moved to the `venumerate()` method of that enumerator. This might be an advantage for longer lists or if the internal enumeration or filtering is otherwise expensive.

ryru commented:
Thank you very much for the review and your time!
